client/go/cmd/test.go

```diff
@@ -0,0 +1,357 @@
+	if statusCode == 0 {
+		statusCode = 200
+	}
+	if statusCode != response.StatusCode {
+		return fmt.Sprintf("Expected status code %d, but got %d, and body:\n%s", statusCode, response.StatusCode, util.ReaderToJSON(response.Body)), nil
+	}
+
+	responseBodySpecBytes, err := getBody(assertion.Response.BodyRaw, testsPath)
+	if err != nil {
+		return "", err
+	}
+	if responseBodySpecBytes == nil {
+		return "", nil
+	}
+	var responseBodySpec interface{}
+	err = json.Unmarshal(responseBodySpecBytes, &responseBodySpec)
+	if err != nil {
+		return "", err
+	}
+
+	responseBodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return "", err
+	}
+	var responseBody interface{}
+	err = json.Unmarshal(responseBodyBytes, &responseBody)
+	if err != nil {
+		return "", err
+	}
+
+	return compare(responseBodySpec, responseBody, "")
+}
+
+func compare(expected interface{}, actual interface{}, path string) (string, error) {
+	typeMatch := false
+	valueMatch := false
+	switch u := expected.(type) {
+	case nil:
+		typeMatch = actual == nil
+		valueMatch = actual == nil
+	case bool:
+		v, ok := actual.(bool)
+		typeMatch = ok
+		valueMatch = ok && u == v
+	case float64:
+		v, ok := actual.(float64)
+		typeMatch = ok
+		valueMatch = ok && math.Abs(u-v) < 1e-9
+	case string:
+		v, ok := actual.(string)
+		typeMatch = ok
+		valueMatch = ok && (u == v)
+	case []interface{}:
+		v, ok := actual.([]interface{})
+		typeMatch = ok
+		if ok && len(u) == len(v) {
+			for i, e := range u {
+				result, err := compare(e, v[i], fmt.Sprintf("%s[%d]", path, i))
+				if result != "" || err != nil {
+					return result, err
+				}
+			}
+			valueMatch = true
+		}
+	case map[string]interface{}:
+		v, ok := actual.(map[string]interface{})
+		typeMatch = ok
+		if ok {
+			for n, e := range u {
+				result, err := compare(e, v[n], fmt.Sprintf("%s['%s']", path, n))
+				if result != "" || err != nil {
+					return result, err
+				}
+			}
+			valueMatch = true
+		}
+	default:
+		return "", fmt.Errorf("unexpected expected JSON type for value '%v'", expected)
+	}
+
+	if !(typeMatch && valueMatch) {
+		if path == "" {
+			path = "root"
+		}
+		expectedJson, _ := json.MarshalIndent(expected, "", "  ")
+		actualJson, _ := json.MarshalIndent(actual, "", "  ")
+		mismatched := "Type"
+		if typeMatch {
+			mismatched = "Value"
+		}
+		return fmt.Sprintf("%s mismatch at %s: <<< expected === actual >>>\n<<<\n%s\n===\n%s\n>>>\n", mismatched, path, expectedJson, actualJson), nil
+	}
+	return "", nil
+}
+
+func getParameters(parametersRaw []byte, testsPath string) (map[string]string, error) {
+	if parametersRaw != nil {
+		var parametersPath string
+		if err := json.Unmarshal(parametersRaw, &parametersPath); err == nil {
+			parametersRaw = readFile(path.Join(testsPath, parametersPath))
+		}
+		var parameters map[string]string
+		if err := json.Unmarshal(parametersRaw, &parameters); err != nil {
+			return nil, fmt.Errorf("request parameters must be JSON with only string values: %v", err)
```

Comment:
Nit: Prefer using `%w` as the format specifier for errors as this allows unwrapping.

Reply:
Thanks!